vm/vm.go

```diff
@@ -725,3 +725,65 @@ func ExecFuncBalancedNativeTokens() ExecFunc {
 		return nil
 	}
 }
+
+func checkAddressRestrictions(output iotago.TxEssenceOutput, address iotago.Address) error {
+	if address.CannotReceiveNativeTokens() && len(output.NativeTokenList()) != 0 {
+		return iotago.ErrAddressCannotReceiveNativeTokens
+	}
+
+	if address.CannotReceiveMana() && output.StoredMana() != 0 {
+		return iotago.ErrAddressCannotReceiveMana
+	}
+
+	if address.CannotReceiveOutputsWithTimelockUnlockCondition() && output.UnlockConditionSet().HasTimelockCondition() {
+		return iotago.ErrAddressCannotReceiveTimelockUnlockCondition
+	}
+
+	if address.CannotReceiveOutputsWithExpirationUnlockCondition() && output.UnlockConditionSet().HasExpirationCondition() {
+		return iotago.ErrAddressCannotReceiveExpirationUnlockCondition
+	}
+
+	if address.CannotReceiveOutputsWithStorageDepositReturnUnlockCondition() && output.UnlockConditionSet().HasStorageDepositReturnCondition() {
+		return iotago.ErrAddressCannotReceiveStorageDepositReturnUnlockCondition
+	}
+
+	if address.CannotReceiveAccountOutputs() && output.Type() == iotago.OutputAccount {
+		return iotago.ErrAddressCannotReceiveAccountOutput
+	}
+
+	if address.CannotReceiveNFTOutputs() && output.Type() == iotago.OutputNFT {
+		return iotago.ErrAddressCannotReceiveNFTOutput
+	}
+
+	if address.CannotReceiveDelegationOutputs() && output.Type() == iotago.OutputDelegation {
+		return iotago.ErrAddressCannotReceiveDelegationOutput
+	}
+
+	return nil
+}
+
+// Returns a func that validates that an address' restrictions are adhered to.
+// Does not validate the Return Address in Storage Deposit Return and Expiration UC.
```

Comment:
And where do we actually check that? Can't we add that check here as well?
IMO the transaction should better be invalid instead of locking funds in nimbus. WDYT?

Reply:
Ah I see, this is done in the additional validators?
Maybe we should add a comment here that this is done in the specific validators?